### PR TITLE
Add apiMaxLength & debugMaxLength settings

### PR DESCRIPTION
**Pull request by hardillb**
part of #3577 and #3581

## Description

This adds the option to configure the apiMaxLength and debugMaxLength Node-RED settings from the template.

This only available to Team/Enterprise licensed instances

Requires "API/Debug Length Limits" to be enabled for team types.

Has a paired nr-launcher PR https://github.com/FlowFuse/nr-launcher/pull/217

image

image

Staging/Prod config change Issue: https://github.com/FlowFuse/CloudProject/issues/377

## Related Issue(s)

#3577 and #3581

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: frontend/src/pages/instance/Settings/Editor.vue [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <template>
     <form class="space-y-6">
-        <TemplateSettingsEditor v-model="editable" :editTemplate="false" />
+        <TemplateSettingsEditor v-model="editable" :editTemplate="false" :team="team" />
```
Comment (Steve-Mcl) [inline frontend/src/pages/instance/Settings/Editor.vue:3]: Ben, is there a commit missing from this PR? 

`TemplateSettingsEditor` does not have a `team` prop.

Comment (Steve-Mcl) [reply frontend/src/pages/instance/Settings/Editor.vue:3]: never mind, i am blind

**Comment by hardillb** [comment]
PostgreSQL Test failure is GH/npm network thing, will re-run when able to.

**Comment by hardillb** [comment]
Hold off on this. It needs a nr-launcher version test

File: forge/db/controllers/ProjectTemplate.js [hunk 2]
```diff
@@ -198,6 +198,16 @@ module.exports = {
...
         }
+        if (result.apiMaxLength) {
+            if (!/^[0-9]+[km]?$/.test(result.apiMaxLength)) {
+                throw new Error('Invalid settings.apiMaxLength')
+            }
+        }
+        if (result.debugMaxLenth) {
+            if (!Number.isInteger(result.debugMaxLength)) {
+                throw new Error('Invalid settings.debugMaxLength')
+            }
```
Comment (Steve-Mcl) [inline forge/db/controllers/ProjectTemplate.js:209]:
```suggestion
            if (!Number.isInteger(result.debugMaxLength) || result.debugMaxLength < 0) {
                throw new Error('Invalid settings.debugMaxLength')
            }
```

Should prevent invalid negative numbers at API level

File: forge/db/controllers/ProjectTemplate.js [hunk 3]
```diff
@@ -198,6 +198,16 @@ module.exports = {
         if (typeof result.httpNodeAuth?.pass === 'string' && result.httpNodeAuth.pass.length > 0) {
             result.httpNodeAuth.pass = hash(result.httpNodeAuth.pass)
         }
+        if (result.apiMaxLength) {
+            if (!/^[0-9]+[km]?$/.test(result.apiMaxLength)) {
```
Comment (Steve-Mcl) [inline forge/db/controllers/ProjectTemplate.js:202]: Should we consider an upper limit?

File: forge/db/controllers/ProjectTemplate.js (hunk 2 above)
Comment (Steve-Mcl) [reply forge/db/controllers/ProjectTemplate.js:209]: Should we not consider an upper limit? greater than `10000` is pushing it but allowing unlimited is possibly a good thing

**Review by Steve-Mcl — COMMENTED** [review]
1 suggestion and a couple of queries. 

Also, I note that if the admin switches the feature flag "API/Debug Length Limits" off, existing limits will continue to persist (but hidden) - should we condition inclusion of these 2 settings based on the feature flag setting?

File: forge/db/controllers/ProjectTemplate.js (hunk 3 above)
Comment (hardillb) [reply forge/db/controllers/ProjectTemplate.js:202]: We can't set a sensible upper limit, we don't know how much memory the NR instance has or what might get dropped by any reverse proxy

File: forge/db/controllers/ProjectTemplate.js (hunk 2 above)
Comment (hardillb) [reply forge/db/controllers/ProjectTemplate.js]: This is a performance thing, if a user cripples themselves it's on them.

**Comment by hardillb** [comment]
>Also, I note that if the admin switches the feature flag "API/Debug Length Limits" off, existing limits will continue to persist (but hidden) - should we condition inclusion of these 2 settings based on the feature flag setting?

Won't that add a bunch of DB queries as part of getting the settings? e.g. need to look up the team -> teamType to check the properties each time the project settings are requested.

**Comment by Steve-Mcl** [comment]
> > Also, I note that if the admin switches the feature flag "API/Debug Length Limits" off, existing limits will continue to persist (but hidden) - should we condition inclusion of these 2 settings based on the feature flag setting?
> 
> Won't that add a bunch of DB queries as part of getting the settings? e.g. need to look up the team -> teamType to check the properties each time the project settings are requested.

Fair point. Not certain of the _cost_ but we can always come back to this if deemed necessary.

Approving as is.

**Review by Steve-Mcl — APPROVED** [review]

**Pull request merged**